### PR TITLE
Handle maintenance mode for yggdrasil

### PR DESCRIPTION
**Pull request by SoulKyu**
Add a new maintenance mode : 

This mode should be configured on the cluster configuration.
When a cluster is in maintenance mode : 
- [x] Add a weight of 0 on all endpoint that are on more than 1 cluster (or just don't add it)
- [x] Don't touch any endpoint that are only on the cluster in maintenance mode
- [x] Yggdrasil wont start if all cluster are in a maintenance mode
- [x] Yggdrasil should send WARNING when their 1 or more cluster in maintenance
- [x] Add a metrics that permit to know cluster in maintenance mode
- [x] Add a metrics that provide informations on which ingress belong to which endpoint

### CODE REVIEW

**Comment by SoulKyu** [comment]
@Solvik @Aluxima

File: cmd/root.go [hunk 1]
```diff
@@ -23,10 +23,11 @@ import (
...
 type clusterConfig struct {
-	APIServer string `json:"apiServer"`
-	Ca        string `json:"ca"`
-	Token     string `json:"token"`
-	TokenPath string `json:"tokenPath"`
+	APIServer   string `json:"apiServer"`
+	Ca          string `json:"ca"`
+	Token       string `json:"token"`
+	TokenPath   string `json:"tokenPath"`
+	Maintenance bool   `json:"maitenance"`
```
Comment (Aluxima) [inline cmd/root.go]:
```suggestion
	Maintenance bool   `json:"maintenance"`
```

File: README.md [hunk 2]
```diff
@@ -167,18 +169,29 @@ The list of certificates will be loaded by Yggdrasil and served to the Envoy nod
...
 Each cluster represents a different Kubernetes cluster with the token being a service account token for that cluster. `ca` is the Path to the ca certificate for that cluster.
 
+Maintenance is a new mode that allow to set a cluster in maintenance mode :
+- Upstream only in one cluster are keeped
+- Upstream in at least 1 cluster that is not in maintenance is deleted for the cluster in maintenance mode
+- Yggdrasil will Fatal if all clusters are in maintenance mode.
+
+This is optional and equal to `false` by default.
+
+kubernetesClusterName is the name of the cluster, its only for information and will be used for metrics. Optional default to `""`
```
Comment (Aluxima) [inline README.md]:
```suggestion
kubernetesClusterName is the name of the cluster, it's only for information and will be used for metrics. Optional defaults to `""`
```

File: README.md [hunk 3]
```diff
@@ -167,18 +169,29 @@ The list of certificates will be loaded by Yggdrasil and served to the Envoy nod
...
-| yggdrasil_listener_updates  | Number of times the listener has been updated  | counter  |
-| yggdrasil_virtual_hosts     | Total number of virtual hosts generated        | gauge    |
+| Name                                         | Description                                    | Type     |
+|----------------------------------------------|------------------------------------------------|----------|
+| yggdrasil_cluster_updates                    | Number of times the clusters have been updated | counter  |
+| yggdrasil_clusters                           | Total number of clusters generated             | gauge    |
+| yggdrasil_ingresses                          | Total number of matching ingress objects       | gauge    |
+| yggdrasil_listener_updates                   | Number of times the listener has been updated  | counter  |
+| yggdrasil_virtual_hosts                      | Total number of virtual hosts generated        | gauge    |
+| yggdrasil_kubernetes_clusters_in_maintenance | Return 1 if cluster in maintenance or 0        | gauge    |
```
Comment (Aluxima) [inline README.md]:
```suggestion
| yggdrasil_kubernetes_cluster_in_maintenance | Return 1 if cluster in maintenance or 0        | gauge    |
```
I would put that in singular as metrics concern each cluster instance

File: pkg/envoy/ingress_translator.go [hunk 4]
```diff
@@ -368,90 +368,130 @@ func validateSubdomain(ruleHost, host string) bool {
...
+			if !ingress.Maintenance {
+				hasNonMaintenance = true
+				break
+			}
+		}
+
+		// Add upstreams based on maintenance status
+		for _, ingress := range ingressList {
+			for _, j := range ingress.Upstreams {
+				if !hasNonMaintenance || !ingress.Maintenance {
```
Comment (Aluxima) [inline pkg/envoy/ingress_translator.go:410]:
```suggestion
				// Skip this upstream if cluster is in maintenance but keep it if no other cluster can serve it 
				if !hasNonMaintenance || !ingress.Maintenance {
```

File: pkg/envoy/ingress_translator.go [hunk 5]
```diff
@@ -368,90 +368,130 @@ func validateSubdomain(ruleHost, host string) bool {
...
+							EnvoyUpstreamInfo.WithLabelValues(strings.ReplaceAll(ruleHost, ".", "_"), j, ingress.Namespace, class, ingress.KubernetesClusterName, ingress.Name).Set(float64(1))
 						}
 					} else {
-						logrus.Warnf("Be careful, healthcheck can't work for wildcard host : %s", envoyIngress.cluster.HealthCheckHost)
+						envoyIngress.addUpstream(j, 1)
+						EnvoyUpstreamInfo.WithLabelValues(strings.ReplaceAll(ruleHost, ".", "_"), j, ingress.Namespace, class, ingress.KubernetesClusterName, ingress.Name).Set(float64(1))
 					}
+				} else {
+					logrus.Warnf("Endpoint is in maintenance mode, upstream %s will not be added for host %s", j, ruleHost)
+					//envoyIngress.addUpstream(j, 0)
```
Comment (Aluxima) [inline pkg/envoy/ingress_translator.go]:
```suggestion
```

File: pkg/envoy/metrics.go [hunk 6]
```diff
@@ -42,8 +42,26 @@ var (
 			Help:      "Number of times the listener has been updated",
 		},
 	)
+
+	KubernetesClusterInMaintenance = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "yggdrasil",
+			Name:      "kubernetes_clusters_in_maintenance",
```
Comment (Aluxima) [inline pkg/envoy/metrics.go]:
```suggestion
			Name:      "kubernetes_cluster_in_maintenance",
```

**Review by Aluxima — CHANGES_REQUESTED** [review]

**Review by SoulKyu — COMMENTED** [review]
Ty

**Review by Aluxima — APPROVED** [review]

**Pull request merged**